zstd-safe/src/seekable.rs

```diff
@@ -0,0 +1,710 @@
+    /// frame and start a new one.
+    pub fn end_frame<C: WriteBuf + ?Sized>(
+        &mut self,
+        output: &mut OutBuffer<'_, C>,
+    ) -> SafeResult {
+        let mut output = output.wrap();
+        // Safety: Just FFI
+        let code = unsafe {
+            zstd_sys::ZSTD_seekable_endFrame(
+                self.0.as_ptr(),
+                ptr_mut(&mut output),
+            )
+        };
+        parse_code(code)
+    }
+
+    /// Wraps the `ZSTD_seekable_endStream()` function. This will end the current frame, and
+    /// then write the seek table so that decompressors can efficiently find compressed frames.
+    ///
+    /// Returns a number > 0 if it was unable to flush all the necessary data to `output`.
+    /// In this case, it should be called again until all remaining data is flushed out and
+    /// 0 is returned.
+    pub fn end_stream<C: WriteBuf + ?Sized>(
+        &mut self,
+        output: &mut OutBuffer<'_, C>,
+    ) -> SafeResult {
+        let mut output = output.wrap();
+        // Safety: Just FFI
+        let code = unsafe {
+            zstd_sys::ZSTD_seekable_endStream(
+                self.0.as_ptr(),
+                ptr_mut(&mut output),
+            )
+        };
+        parse_code(code)
+    }
+}
+
+impl<'a> Drop for SeekableCStream<'a> {
+    fn drop(&mut self) {
+        // Safety: Just FFI
+        unsafe {
+            zstd_sys::ZSTD_seekable_freeCStream(self.0.as_ptr());
+        }
+    }
+}
+
+/// Allows for the seek table to be constructed directly. This table can then be appended
+/// to a file of concatenated frames. This allows the frames to be compressed independently,
+/// even in parallel, and compiled together afterward into a seekable archive.
+pub struct FrameLog<'a>(NonNull<zstd_sys::ZSTD_frameLog>, PhantomData<&'a ()>);
+
+unsafe impl Send for FrameLog<'_> {}
+unsafe impl Sync for FrameLog<'_> {}
+
+impl<'a> FrameLog<'a> {
+    /// Tries to create a new `FrameLog`.
+    ///
+    /// Returns `None` if zstd returns a NULL pointer - may happen if allocation fails.
+    pub fn try_create(checksum_flag: i32) -> Option<Self> {
+        Some(FrameLog(
+            // Safety: Just FFI
+            NonNull::new(unsafe {
+                zstd_sys::ZSTD_seekable_createFrameLog(checksum_flag)
+            })?,
+            PhantomData,
+        ))
+    }
+
+    /// Creates a new `FrameLog`.
+    ///
+    /// # Panics
+    ///
+    /// If zstd returns a NULL pointer.
+    pub fn create(checksum_flag: i32) -> Self {
+        Self::try_create(checksum_flag)
+            .expect("Zstd returned null pointer when creating new frame log")
+    }
+
+    /// Needs to be called once for each frame in the archive. `checksum` is optional, and
+    /// will not be used if `checksum_flag` was 0 when the frame log was created. If present,
+    /// it should be the least significant 32 bits of the XXH64 hash of the uncompressed data.
```

Comment:
`checksum` needs to be here if `checksumFlag` was enabled. The original wording (which is copied here) is a bit ambiguous, and may imply `checksum` is optional, even if  `checksumFlag` was enabled. Could we make it very clear that if `checksumFlag` was enabled, then `checksum` is required, and that it's ignored otherwise?

I was wondering if we could somehow take an `Option` to clearly indicate the fact that it's optional, but I don't see a clean way of doing that without making `checksumFlag` some const generic parameter (which we don't want).

Reply:
I made it an option to reflect that it's optional and updated the description.